os2phonebook_service/os2phonebook/datastore.py

```diff
@@ -0,0 +1,622 @@
+from typing import List, Tuple, Callable
+from elasticsearch import Elasticsearch
+from os2phonebook.exceptions import InvalidSearchType
+
+
+def create_connection(host: str, port: int) -> Elasticsearch:
+    """Elasticsearch connection factory.
+
+    This is a connection factory to connect to the
+    Elasticsearch backend
+
+    Args:
+        host (str): Hostname of the backend
+        port (int): Service port
+
+    Returns:
+        :obj:`Elasticsearch`: An instance of the elasticsearch client
+
+    """
+    if not isinstance(host, str):
+        raise TypeError("Host description must be passed as a string")
+
+    if not isinstance(port, int):
+        raise TypeError("Port identifier must be passed as an integer")
+
+    db = Elasticsearch([
+        {
+            'host': host,
+            'port': port
+        },
+    ])
+
+    return db
+
+
+class DataStore(object):
+    """Client for generating queries in backend datastore
+
+    Args:
+        db (:obj:`Elasticsearch`): Instance of the elasticsearch client.
+
+        search_type_map (dict): A map of all the available search types.
+
+            Available search types:
+                * employee_by_name
+                * employee_by_phone
+                * employee_by_email
+                * employee_by_engagement
+                * org_unit_by_name
+
+            Example:
+
+            {
+                "employee_by_name": {
+                    "description": "Navn",
+                    "query_method": "query_for_employee_by_name"
+            }
+
+            This particular search type has the canonical description `Navn`
+            and a query method which refers to the name of an attribute
+            on this class, e.g. `Datastore.query_for_employee_by_name`.
+
+    """
+
+    def __init__(self, db):
+
+        if not isinstance(db, Elasticsearch):
+            raise TypeError("Datastore requires an instance of the Elasticsearch object")
+
+        self.db = db
+
+        self.search_type_map =  {
+            "employee_by_name": {
+                "description": "Navn",
+                "query_method": "query_for_employee_by_name"
+            },
+            "employee_by_phone": {
+                "description": "Telefon",
+                "query_method": "query_for_employee_by_phone"
+            },
+            "employee_by_email": {
+                "description": "Email",
+                "query_method": "query_for_employee_by_email"
+            },
+            "employee_by_engagement": {
+                "description": "Stilling",
+                "query_method": "query_for_employee_by_engagement"
+            },
+            "org_unit_by_name": {
+                "description": "Enhed",
+                "query_method": "query_for_org_unit_by_name"
+            }
+        }
```

Comment:
I think you could have the actual functions here as the query methods, instead of having to do the lookup from the strings.